projects/plugins/jetpack/modules/shortcodes/smartframe.php

```diff
@@ -0,0 +1,124 @@
+
+	// Actually display the smartframe Embed.
+	add_shortcode( 'smartframe', 'jetpack_smartframe_shortcode' );
+}
+
+/**
+ * Compose shortcode based on smartframe iframes.
+ *
+ * @since 9.3.3
+ *
+ * @param string $content Post content.
+ *
+ * @return mixed
+ */
+function wpcom_shortcodereverse_smartframe( $content ) {
+	if ( ! is_string( $content ) || false === stripos( $content, 'embed.smartframe' ) ) {
+		return $content;
+	}
+
+	// phpcs:ignore WordPress.WP.EnqueuedResources.NonEnqueuedScript
+	$regexp     = '!<script\ssrc="https://embed\.smartframe\.(io|net)/(\w+)\.js"\sdata-image-id="(.*?)"(\sdata-width="(\d+(%|px))"\s)?(data-max-width="(\d+(%|px)))?"></script>!i';
+	$regexp_ent = str_replace( '&amp;#0*58;', '&amp;#0*58;|&#0*58;', htmlspecialchars( $regexp, ENT_NOQUOTES ) );
+
+	foreach ( compact( 'regexp', 'regexp_ent' ) as $regexp ) {
+		if ( ! preg_match_all( $regexp, $content, $matches, PREG_SET_ORDER ) ) {
+			continue;
+		}
+
+		foreach ( $matches as $match ) {
+			$script_id = $match[2];
+
+			if ( empty( $script_id ) ) {
+				continue;
+			}
+
+			$image_id = $match[3];
+
+			if ( empty( $image_id ) ) {
+				continue;
+			}
+
+			if ( isset( $match[5] ) ) {
+				$width = $match[5];
+			}
+
+			if ( isset( $match[8] ) ) {
+				$max_width = $match[8];
+			}
+
+			$shortcode = '[smartframe script-id="' . $script_id . '" image-id="' . esc_attr( $image_id ) . '"';
+			if ( ! empty( $width ) ) {
+				$shortcode .= ' width="' . esc_attr( $width ) . '"';
+			}
+			if ( ! empty( $max_width ) ) {
+				$shortcode .= ' max-width="' . esc_attr( $max_width ) . '"';
+			}
+			$shortcode .= ']';
+
+			$content = str_replace( $match[0], $shortcode, $content );
+		}
+	}
+	/** This action is documented in modules/widgets/social-media-icons.php */
+	do_action( 'jetpack_bump_stats_extras', 'html_to_shortcode', 'smartframe' );
+
+	return $content;
+}
+
+/**
+ * Parse shortcode arguments and render its output.
+ *
+ * @since 9.3.3
+ *
+ * @param array $atts Shortcode parameters.
+ *
+ * @return string
+ */
+function jetpack_smartframe_shortcode( $atts ) {
+	if ( ! empty( $atts['image-id'] ) ) {
+		$image_id = $atts['image-id'];
+	} else {
+		return '<!-- Missing smartframe image-id -->';
+	}
+	if ( ! empty( $atts['script-id'] ) ) {
+		$script_id = $atts['script-id'];
+	} else {
+		return '<!-- Missing smartframe script-id -->';
+	}
+
+	$params = array(
+		// ignore width for now, smartframe embed code has it "100%". % isn't allowed in oembed, making it 100px.
+		// 'width'  => isset( $atts['width'] ) ? (int) $atts['width'] : null,.
+		'max-width' => isset( $atts['max-width'] ) ? (int) $atts['max-width'] : null,
+	);
+
+	// wrap the embed with wp-block-embed__wrapper, otherwise it would be aligned to the very left of the viewport.
+	return "<div class='wp-block-embed__wrapper'>" . wp_oembed_get( 'https://imagecards.smartframe.io/' . $script_id . '/' . $image_id, array_filter( $params ) ) . '</div>';
```

Comment:
Would it be more readable to do something like this?

```suggestion
	$embed_url = sprintf(
		'https://imagecards.smartframe.io/%1$s/%2$s',
		esc_attr( $script_id ),
		esc_attr( $image_id )
	);

	// wrap the embed with wp-block-embed__wrapper, otherwise it would be aligned to the very left of the viewport.
	return sprintf(
		'<div class="wp-block-embed__wrapper">%1$s</div>',
		wp_oembed_get( $embed_url, array_filter( $params ) )
	);
```